### ISSUE TITLE
Change Snakemake rules after merge to depend on compressed file names (.gz) ?

### ISSUE THREAD

**Issue by kvarforl** [RTXteam/RTX#1104]
I noticed while debugging other things with the Snakemake Finish rule that using the `-R Finish` flag to resume a build that failed in the Finish rule re computes every rule from Merge down. This is the dryrun output from this circumstance:
```
139 Job counts:
140         count   jobs
141         1       Finish
142         1       Merge
143         1       Nodes
144         1       Simplify
145         1       Simplify_Nodes
146         1       Simplify_Stats
147         1       Slim
148         1       Stats
149         1       TSV
150         9
151 This was a dry-run (flag -n). The order of jobs does not reflect the order of execution.
```
I think this is because many of the input files the Finish rule expects have been compressed, so their filename ends in ".gz", which causes snakemake to not recognize their existence. 

In an effort ot verify this, I ran `touch kg2-build/kg2-nodes.json kg2-build/kg2-orphans-edges.json kg2-build/kg2-simplified-nodes.json kg2-build/kg2-simplified.json kg2-build/kg2-slim.json`  and then did a dryrun of the snakemake build again. (i removed the empty files afterwards to avoid confusion). this is the job output from this circumstance:
```
 91 Job counts:
 92         count   jobs
 93         1       Finish
 94         1       Simplify_Stats
 95         1       TSV
 96         3
 97 This was a dry-run (flag -n). The order of jobs does not reflect the order of execution.
```

In summary, the zipping does seem to impact snakemake's perception of which jobs need to be executed, but more investigation needs to be done as to why the Simplify_Stats and TSV rules are still expected to run when their ouput files do seem to exist.

**Comment by saramsey** [comment]
Thanks for the update. Would it make sense to document this workaround in the README.md somewhere?

**Comment by saramsey** [comment]
One possibility would be to change these lines (lines 46 and 50--54) of `finish-snakemake.sh`
```
gzip -f ${final_output_file_full}
tar -C ${kg2_tsv_dir} -czvf ${kg2_tsv_dir} nodes.tsv nodes_header.tsv edges.tsv edges_header.tsv
${s3_cp_cmd} ${kg2_tsv_tarball} s3://${s3_bucket}/

gzip -f ${simplified_output_file_full}
gzip -f ${simplified_output_nodes_file_full}
gzip -f ${output_nodes_file_full}
gzip -f ${output_file_orphan_edges}
gzip -f ${slim_output_file_full}
```
so that they do not delete the original files. See 
https://unix.stackexchange.com/questions/46786/how-to-tell-gzip-to-keep-original-file/58814

**Comment by kvarforl** [comment]
That makes a lot of sense!
The current workaround that I have running right now is changing lines 3-10 in `Snakefile-finish`, which caused snakemake to only run the rule Finish (executing now). 
```
  1 rule Finish:
  2     input:
  3         final_output_file_full = config['FINAL_OUTPUT_FILE_FULL']+ ".gz",
  4         output_file_orphan_edges = config['OUTPUT_FILE_ORPHAN_EDGES']+ ".gz",
  5         report_file_full = config['REPORT_FILE_FULL'],
  6         output_nodes_file_full = config['OUTPUT_NODES_FILE_FULL']+ ".gz",
  7         simplified_output_file_full = config['SIMPLIFIED_OUTPUT_FILE_FULL']+ ".gz",
  8         simplified_report_file_full = config['SIMPLIFIED_REPORT_FILE_FULL'],
  9         simplified_output_nodes_file_full = config['SIMPLIFIED_OUTPUT_NODES_FILE_FULL']+ ".gz",
 10         slim_output_file_full = config['SLIM_OUTPUT_FILE_FULL']+ ".gz",
 11         placeholder = config['BUILD_DIR'] + "/tsv_placeholder.empty"
```

I think I like your solution better; it seems like it will be less confusing to maintain, less likely to mess with other snakemake rules, and require fewer changes to the code overall :) Thanks for the suggestion!

**Comment by kvarforl** [comment]
@saramsey  I'm working on writing the script to delete the big files at the end of the build, but I'm having some struggles deciding the best way to get the names of the files to be deleted to the script.
it doesn't make much sense to pass them in as command line arguments since the script will be run independently of the build, and hardcoding them seems a bit dicey and repetitive as well. 

Should I add them to `master_config.shinc`? 
Alternatively, I could create a separate .shinc file at the end of `finish-snakemake.sh` with the variables and names that were passed to it; that way they should be consistent with the names of the files used in the most recent build.

Do you have a preference for either method?

**Comment by saramsey** [comment]
That shell script can be called `post-build-cleanup.sh` and it can look like this (needs header comments, etc., of course):
```
#!/bin/bash
set -o nounset -o pipefail -o errexit
config_dir=`dirname "$0"`
source ${config_dir}/master-config.shinc
rm -f ${build_dir}/kg2.json
rm -f ${build_dir}/kg2-nodes.json
rm -f ${build_dir}/kg2-simplified.json
rm -f ${build_dir}/kg2-simplified-nodes.json
...

**Comment by kvarforl** [comment]
Okay, so hard coding the filenames is a nonissue? 

Thanks for the feedback!

**Comment by saramsey** [comment]
OK, good point!  What about just adding `rm` commands to the end of `finish-snakemake.sh`?  Then you have shell variables for each filename? Does that make sense?

**Comment by kvarforl** [comment]
That would make sense, but wouldn't fully solve the problem of doing partial rebuilds. If we run into a circumstance (like the one we're in right now, fixing mappings that get implemented in Simplify) where the build successfully completed but we want to change something in between Merge and Finish, we'd have the same problem of the Snakefile not recognizing the uncompressed files (because they would have already been deleted).

Does that make sense?

**Comment by kvarforl** [comment]
I added the script with the hardcoded file names. I'll adjust it when [#1114](https://github.com/RTXteam/RTX/issues/1114) is addressed :) Closing for now
